broken-line/contracts/LibBrokenLine.sol

```diff
@@ -5,55 +5,109 @@ pragma abicoder v2;
+  * because users can only run out of lockup periods.
   **/
-//todo подумать про типы, нужна ли максимальная точность
-contract BrokenLineDomain {
+
+library LibBrokenLine {
+    using SignedSafeMathUpgradeable for int;
+    using SafeMathUpgradeable for uint;
+    using LibIntMapping for mapping (uint => int);
+
     struct Line {
         uint start;
         uint bias;
         uint slope;
     }
 
+    struct LineData {       //all data about line
+        Line line;
+        uint cliff;
+    }
+
     struct BrokenLine {
-        mapping (uint => int) slopeChanges; //change of slope applies to the next time point
+        mapping (uint => int) slopeChanges;         //change of slope applies to the next time point
+        mapping (uint => LineData) initiatedLines;  //initiated (successfully added) Lines
         Line initial;
     }
-}
 
-library LibBrokenLine {
-    using SignedSafeMathUpgradeable for int;
-    using SafeMathUpgradeable for uint;
+    /**
+     *add Line, save data in LineData
+     **/
+    function add(BrokenLine storage brokenLine, uint id, Line memory line, uint cliff) internal {
+        require(line.slope != 0, "Slope == 0, unacceptable value for slope");
+        require(line.slope <= line.bias, "Slope > bias, unacceptable value for slope");
+        require(brokenLine.initiatedLines[id].line.bias == 0, "Line with given id is already exist");
+        brokenLine.initiatedLines[id] = LineData(line, cliff);
 
-    function add(BrokenLineDomain.BrokenLine storage brokenLine, BrokenLineDomain.Line memory line, uint cliff) internal {
         update(brokenLine, line.start);
         brokenLine.initial.bias = brokenLine.initial.bias.add(line.bias);
         uint period = line.bias.div(line.slope);
         if (cliff == 0) {
             brokenLine.initial.slope = brokenLine.initial.slope.add(line.slope);
         } else {
             uint cliffEnd = line.start.add(cliff).sub(1);
-            brokenLine.slopeChanges[cliffEnd] = brokenLine.slopeChanges[cliffEnd].add(safeInt(line.slope));
+            brokenLine.slopeChanges.addToItem(cliffEnd, safeInt(line.slope));
             period = period.add(cliff);
         }
 
         int mod = safeInt(line.bias.mod(line.slope));
         uint256 endPeriod = line.start.add(period);
         uint256 endPeriodMinus1 = endPeriod.sub(1);
-        brokenLine.slopeChanges[endPeriodMinus1] = brokenLine.slopeChanges[endPeriodMinus1].sub(safeInt(line.slope)).add(mod);
-        brokenLine.slopeChanges[endPeriod] = brokenLine.slopeChanges[endPeriod].sub(mod);
+        brokenLine.slopeChanges.subFromItem(endPeriodMinus1, safeInt(line.slope).sub(mod));
+        brokenLine.slopeChanges.subFromItem(endPeriod, mod);
+    }
+
+    /**
+     *remove Line from BrokenLine, return line.bias, which actual now moment
+     **/
+    function remove(BrokenLine storage brokenLine, uint id, uint toTime) internal returns (uint) {
+        LineData memory lineData = brokenLine.initiatedLines[id];
```

Comment:
why no require here? do we need to check if line exists?

Reply:
Writing simple test show: removing line, which not exist goes to exception "Reason given: SafeMath: division by zero".
So need to check if line exists

Reply:
Done? add require